cyral/resource_cyral_repository_binding.go

```diff
@@ -2,120 +2,222 @@ package cyral
+				Computed:    true,
 				Type:        schema.TypeString,
+			},
+			SidecarIDKey: {
+				Description: "ID of the sidecar that will be bound to the given repository.",
 				Required:    true,
 				ForceNew:    true,
+				Type:        schema.TypeString,
 			},
-			"repository_id": {
+			RepositoryIDKey: {
 				Description: "ID of the repository that will be bound to the sidecar.",
-				Type:        schema.TypeString,
 				Required:    true,
 				ForceNew:    true,
-			},
-			"listener_port": {
-				Description: "Port in which the sidecar will listen for the given repository.",
-				Type:        schema.TypeInt,
-				Required:    true,
-			},
-			"listener_host": {
-				Description: "Address in which the sidecar will listen for the given repository. By default, the sidecar will listen in all interfaces.",
 				Type:        schema.TypeString,
-				Optional:    true,
-				Default:     "0.0.0.0",
 			},
-			"sidecar_as_idp_access_gateway": {
-				Description: "Indicates whether or not the sidecar in the binding configuration is selected as the Access Gateway for Identity Provider users connecting to the underlying data repository. Defaults to `false`.",
+			BindingEnabledKey: {
+				Description: "Enable or disable all listener bindings.",
 				Type:        schema.TypeBool,
 				Optional:    true,
-				Default:     false,
-			},
-			"id": {
-				Description: "Computed ID for this resource (locally computed to be used in Terraform state).",
-				Computed:    true,
-				Type:        schema.TypeString,
+				Default:     true,
 			},
-		},
-	}
-}
-
-// The upgrade from v0 to v1 of the repository binding resource consists simply
-// in changing the format of the `id` from `{sidecar_id}-{repository_id}` to
-// `{sidecar_id}/{repository_id}`.
-func upgradeRepositoryBindingV0(
-	_ context.Context,
-	rawState map[string]interface{},
-	meta interface{},
-) (map[string]interface{}, error) {
-	prevID := rawState["id"].(string)
-	prevSep := "-"
-	newSep := "/"
-	ids, err := unmarshalComposedID(prevID, prevSep, 2)
-	if err != nil {
-		// If we ignore this error, the ID will be inconsistent with
-		// what we expect in v1. We should not let that happen,
-		// therefore return error here.
-		return rawState, fmt.Errorf("unable to unmarshal composed ID: %w", err)
-	}
-	sidecarID := ids[0]
-	repositoryID := ids[1]
-	rawState["id"] = marshalComposedID([]string{sidecarID, repositoryID}, newSep)
-	return rawState, nil
-}
-
-func resourceRepositoryBinding() *schema.Resource {
-	return &schema.Resource{
-		Description:   "Manages [repositories to sidecars binding](https://cyral.com/docs/sidecars/sidecar-assign-repo).",
-		CreateContext: resourceRepositoryBindingCreate,
-		ReadContext:   resourceRepositoryBindingRead,
-		UpdateContext: resourceRepositoryBindingUpdate,
-		DeleteContext: resourceRepositoryBindingDelete,
-
-		SchemaVersion: 1,
-		StateUpgraders: []schema.StateUpgrader{
```

Comment:
Good point, I wasn't sure how to handle that. I can re-add the old state upgraders-- but I figured there is no point, since the migration script will need to be run anyway (in order to upgrade to 4.0, all bindings will be migrated in the CP and then imported into terraform, like the previous migration). I investigated using the state upgraders for migration, however I decided against it because I was concerned with the fact that if the StateUpgrade fails for some reason, it would be hard to catch and address it (since state upgrade happens automatically when terraform apply is run). Additionally, since bindings are migrated in the CP using various validations, I figured it was safer to just import those back into terraform to eliminate potential inconsistencies. Also, since listeners need to be created based on bindings, there needs to be a script anyway.